### PR TITLE
Fix duplicate username registration v2

### PR DESCRIPTION
**Pull request by leblowl**
Another attempt at refactoring the registration service so that we can more easily verify that it does not register duplicate certs. Related to #2155 

### Pull Request Checklist

- [x] I have linked this PR to related GitHub issue.
- [x] I have updated the CHANGELOG.md file with relevant changes (the file is located at the root of monorepo).

### (Optional) Mobile checklist

Please ensure you completed the following checks if you did any changes to the mobile package:

- [ ] I have run e2e tests for mobile
- [ ] I have updated base screenshots for visual regression tests

### CODE REVIEW

File: packages/backend/src/nest/storage/certificates/certificates.store.ts [hunk 1]
```diff
@@ -97,6 +92,11 @@ export class CertificatesStore {
   public updateMetadata(metadata: CommunityMetadata) {
     if (!metadata) return
     this.metadata = metadata
+    // Community metadata is required for validating certificates, so
+    // we re-validate certificates once community metadata is set. It
+    // might also make sense to initialize CertificateStore after
+    // CommunityMetadata is received instead of doing this.
+    this.loadedCertificates()
```
Comment (leblowl) [inline packages/backend/src/nest/storage/certificates/certificates.store.ts]: The previous wait mechanism with setTimeout appeared to be causing another JS timing issue where callbacks were run out of order and caused some certificates to not be sent to the frontend. The other changes in this PR seemed to flush the bug out into the open. So this change is fixing that. Now instead of waiting, we throw an error if community metadata is missing and revalidate when community metadata is available.

File: packages/backend/src/nest/registration/registration.service.ts [hunk 2]
```diff
@@ -4,35 +4,66 @@ import { extractPendingCsrs, issueCertificate } from './registration.functions'
...
+        this.logger('Issuing certificates', event)
+        this.registrationEventInProgress = true
+        await this.issueCertificates({
+          ...event,
+          certificates: (await this.storageService?.certificatesStore.loadAllCertificates()) as string[],
+        })
       }
-    )
+    }
   }
```
Comment (leblowl) [inline packages/backend/src/nest/registration/registration.service.ts:46]: Process only one event at a time (`registrationEventInProgress`).

File: packages/backend/src/nest/registration/registration.service.ts [hunk 3]
```diff
@@ -75,10 +99,11 @@ export class RegistrationService extends EventEmitter implements OnModuleInit {
...
 
   public async registerUserCertificate(csr: string): Promise<void> {
-    const result = await issueCertificate(csr, this._permsData)
+    const result = await issueCertificate(csr, this.permsData)
     this.logger('DuplicatedCertBug', { result })
     if (result?.cert) {
-      this.emit(RegistrationEvents.NEW_USER, { certificate: result.cert })
+      // @ts-ignore
+      await this.storageService?.saveCertificate({ certificate: result.cert })
```
Comment (leblowl) [inline packages/backend/src/nest/registration/registration.service.ts:106]: We only process the next certificate after we have saved the current batch of certificates and we know we have saved them because we are awaiting it here rather than sending events which are not awaited.

File: packages/backend/src/nest/connections-manager/connections-manager.service.ts [hunk 4]
```diff
@@ -618,15 +613,12 @@ export class ConnectionsManagerService extends EventEmitter implements OnModuleI
...
-        this.serverIoProvider.io.emit(SocketActionTypes.RESPONSE_GET_CSRS, { csrs: payload.csrs })
-        this.registrationService.emit(RegistrationEvents.REGISTER_USER_CERTIFICATE, payload)
-      }
-    )
+    this.storageService.on(StorageEvents.REPLICATED_CSR, async (payload: { csrs: string[] }) => {
+      this.logger(`Storage - ${StorageEvents.REPLICATED_CSR}`)
+      this.libp2pService.emit(Libp2pEvents.DIAL_PEERS, await getLibp2pAddressesFromCsrs(payload.csrs))
+      this.serverIoProvider.io.emit(SocketActionTypes.RESPONSE_GET_CSRS, payload)
+      this.registrationService.emit(RegistrationEvents.REGISTER_USER_CERTIFICATE, payload)
+    })
```
Comment (leblowl) [inline packages/backend/src/nest/connections-manager/connections-manager.service.ts:631]: Just a small refactor and lint

File: packages/backend/src/nest/storage/certificates/certificates.store.ts [hunk 5]
```diff
@@ -54,27 +50,26 @@ export class CertificatesStore {
...
+    await this.loadedCertificates()
 
     this.logger('Initialized')
   }
 
+  public async loadedCertificates() {
+    this.emitter.emit(StorageEvents.LOADED_CERTIFICATES, {
+      certificates: await this.getCertificates(),
+    })
+  }
```
Comment (leblowl) [inline packages/backend/src/nest/storage/certificates/certificates.store.ts:72]: Just moving this outside the init function so it can be used below

File: packages/backend/src/nest/storage/storage.service.ts [hunk 6]
```diff
@@ -72,7 +72,7 @@ export class StorageService extends EventEmitter {
     private readonly localDbService: LocalDbService,
     private readonly orbitDbService: OrbitDb,
     private readonly certificatesRequestsStore: CertificatesRequestsStore,
-    private readonly certificatesStore: CertificatesStore,
+    public readonly certificatesStore: CertificatesStore,
```
Comment (leblowl) [inline packages/backend/src/nest/storage/storage.service.ts]: Making this public so we can access it from the registration service. We can also build wrapper functions in storage service and keep it private.

File: packages/backend/src/nest/storage/storage.service.ts [hunk 7]
```diff
@@ -759,11 +754,6 @@ export class StorageService extends EventEmitter {
     })
   }
 
-  public resetCsrAndCertsValues() {
-    this.certificatesRequestsStore.resetCsrReplicatedMapAndId()
-    this.certificatesStore.resetValues()
-  }
```
Comment (leblowl) [inline packages/backend/src/nest/storage/storage.service.ts:782]: Is there any need to call `this.certificatesStore.resetValues()`?

Comment (EmiM) [reply packages/backend/src/nest/storage/storage.service.ts:782]: I think it's needed for restarting Quiet. We do not close backend completely when user leaves community, we just purge data and clean the state, reuse the tor process.

File: packages/backend/src/nest/storage/storage.service.ts (hunk 6 above)
Comment (EmiM) [reply packages/backend/src/nest/storage/storage.service.ts]: Yeah, I'd prefer to add accessors if you don't mind.

Comment (leblowl) [reply packages/backend/src/nest/storage/storage.service.ts]: Like getters and setters? Or wrapper functions for specific CertificatesStore functions, e.g. `StorageService.getCertificates` which is a wrapper for `CertificatesStore.getCertificates`?

Comment (EmiM) [reply packages/backend/src/nest/storage/storage.service.ts]: Either way is fine for me but I think getter/setter is a slightly better choice in this case.

File: packages/backend/src/nest/registration/registration.service.ts [hunk 8]
```diff
@@ -4,35 +4,75 @@ import { extractPendingCsrs, issueCertificate } from './registration.functions'
...
 import { RegistrationEvents } from './registration.types'
 import Logger from '../common/logger'
+import { StorageService } from '../storage/storage.service'
 
 @Injectable()
 export class RegistrationService extends EventEmitter implements OnModuleInit {
   private readonly logger = Logger(RegistrationService.name)
   public certificates: string[] = []
-  private _permsData: PermsData
+  public permsData: PermsData
```
Comment (EmiM) [inline packages/backend/src/nest/registration/registration.service.ts]: Why is this public? If it's only for setting permsData in test, maybe let's add a setter instead? Or something simliar

**Review by EmiM — CHANGES_REQUESTED** [review]
Just a few comments. I'd also look at e2e tests because none of them passed.

File: packages/backend/src/nest/registration/registration.service.ts (hunk 8 above)
Comment (leblowl) [reply packages/backend/src/nest/registration/registration.service.ts]: I'm happy to change it back to using a setter as it was before. It's a very minor thing. But just to explain why I changed it: I don't see much benefit of a setter/getter if it's not doing anything beyond setting/getting the value. That's what public is for and doesn't require a variable name like `_permsData`. From a caller's perspective the syntax of a setter and public are the same. If we want to avoid public variables for some reason, then I'd probably prefer using methods like `setPermsData` over JS's setter/getter functionality.

File: packages/backend/src/nest/storage/storage.service.ts (hunk 6 above)
Comment (leblowl) [reply packages/backend/src/nest/storage/storage.service.ts]: I went with the wrapper here because that approach was already being used in this file

File: packages/backend/src/nest/storage/storage.service.ts (hunk 7 above)
Comment (leblowl) [reply packages/backend/src/nest/storage/storage.service.ts:782]: I've added clean methods back in and added them to stores that didn't have them. So the code should be pretty similar to how it was before we refactored things (when everything was in the storage service).

**Review by EmiM — APPROVED** [review]
